From 270b461e88b47781b3ef9bba2779074d0eb4996b Mon Sep 17 00:00:00 2001
From: Waldemar Brodkorb <wbx@openadk.org>
Date: Fri, 25 Aug 2023 05:04:53 +0200
Subject: [PATCH] .rofixup fix

Signed-off-by: Waldemar Brodkorb <wbx@openadk.org>
---
 elf2flt.ld.in | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/elf2flt.ld.in b/elf2flt.ld.in
index c7e01a6..c69531e 100644
--- a/elf2flt.ld.in
+++ b/elf2flt.ld.in
@@ -31,6 +31,7 @@ W_RODAT		*(.rodata)
 W_RODAT		*(.rodata1)
 W_RODAT		*(.rodata.*)
 W_RODAT		*(.gnu.linkonce.r*)
+W_RODAT		*(.rofixup)
 
 		/* This is special code area at the end of the normal
 		   text section.  It contains a small lookup table at
@@ -60,6 +61,7 @@ R_RODAT		*(.rodata)
 R_RODAT		*(.rodata1)
 R_RODAT		*(.rodata.*)
 R_RODAT		*(.gnu.linkonce.r*)
+R_RODAT		*(.rofixup)
 		*(.data)
 		*(.data1)
 		*(.data.*)
-- 
2.30.2

