diff -Nur busybox-1.25.0.orig/networking/ifupdown.c busybox-1.25.0/networking/ifupdown.c
--- busybox-1.38.0.orig/networking/ifupdown.c	2026-04-17 17:44:06.000000000 +0200
+++ busybox-1.38.0/networking/ifupdown.c	2026-08-27 12:32:06.228513824 +0200
@@ -636,7 +636,7 @@ static const struct dhcp_client_t ext_dh
 	},
 	{ "udhcpc",
 		"udhcpc " UDHCPC_CMD_OPTIONS " -p /var/run/udhcpc.%iface%.pid -i %iface%[[ -x hostname:%hostname%]][[ -c %client%]]"
-				"[[ -s %script%]][[ %udhcpc_opts%]]",
+				"[[ -s %script%]][[ %udhcpc_opts%]] 2>/dev/null",
 		"kill `cat /var/run/udhcpc.%iface%.pid` 2>/dev/null",
 	},
 };
@@ -675,7 +675,7 @@ static int FAST_FUNC dhcp_up(struct inte
 		return 0;
 #  endif
 	return execute("udhcpc " UDHCPC_CMD_OPTIONS " -p /var/run/udhcpc.%iface%.pid "
-			"-i %iface%[[ -x hostname:%hostname%]][[ -c %client%]][[ -s %script%]][[ %udhcpc_opts%]]",
+			"-i %iface%[[ -x hostname:%hostname%]][[ -c %client%]][[ -s %script%]][[ %udhcpc_opts%]] 2>/dev/null",
 			ifd, exec);
 }
 # else
