--- busybox-1.38.0.orig/include/platform.h	2026-02-06 08:45:29.000000000 +0100
+++ busybox-1.38.0/include/platform.h	2026-08-27 12:32:06.236513825 +0200
@@ -159,6 +159,7 @@
 # include <sex.h>
 #elif defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__NetBSD__) \
    || defined(__APPLE__)
+# include <sys/param.h>		/* contains needed types for the following */
 # include <sys/resource.h>  /* rlimit */
 # include <machine/endian.h>
 # define bswap_64 __bswap64
