diff -Nur busybox-1.19.4.orig/scripts/kconfig/Makefile busybox-1.19.4/scripts/kconfig/Makefile
--- busybox-1.38.0.orig/scripts/kconfig/Makefile	2023-01-05 16:29:10.000000000 +0100
+++ busybox-1.38.0/scripts/kconfig/Makefile	2026-08-27 12:32:06.244513826 +0200
@@ -114,7 +114,7 @@ help:
 #         Based on GTK which needs to be installed to compile it
 # object files used by all kconfig flavours
 
-hostprogs-y	:= conf mconf qconf gconf kxgettext
+hostprogs-y	:= conf qconf gconf kxgettext
 conf-objs	:= conf.o  zconf.tab.o
 mconf-objs	:= mconf.o zconf.tab.o
 kxgettext-objs	:= kxgettext.o zconf.tab.o
