diff -Nur busybox-1.36.1.orig/Makefile.flags busybox-1.36.1/Makefile.flags
--- busybox-1.38.0.orig/Makefile.flags	2024-09-26 16:00:52.000000000 +0200
+++ busybox-1.38.0/Makefile.flags	2026-08-27 12:32:06.248513826 +0200
@@ -88,14 +88,14 @@ endif
 #CFLAGS += $(call cc-option,-Wconversion,)
 
 ifneq ($(CONFIG_DEBUG),y)
-CFLAGS += $(call cc-option,-Oz,$(call cc-option,-Os,$(call cc-option,-O2,)))
+CFLAGS += $(call cc-option,-Os,$(call cc-option,-O2,))
 else
 CFLAGS += $(call cc-option,-g,)
 #CFLAGS += "-D_FORTIFY_SOURCE=2"
 ifeq ($(CONFIG_DEBUG_PESSIMIZE),y)
 CFLAGS += $(call cc-option,-O0,)
 else
-CFLAGS += $(call cc-option,-Oz,$(call cc-option,-Os,$(call cc-option,-O2,)))
+CFLAGS += $(call cc-option,-Os,$(call cc-option,-O2,))
 endif
 endif
 ifeq ($(CONFIG_DEBUG_SANITIZE),y)
