From 0dfba20346541d0a9261f1d5c81f9d4ba5122b26 Mon Sep 17 00:00:00 2001
From: Waldemar Brodkorb <wbx@openadk.org>
Date: Sun, 25 Jan 2026 18:50:49 +0100
Subject: [PATCH] fix gettid declaration

---
 include/unistd.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/unistd.h b/include/unistd.h
index 0fd430549..bfb6ecc13 100644
--- a/include/unistd.h
+++ b/include/unistd.h
@@ -1260,7 +1260,7 @@ extern int getentropy(void *__buf, size_t __len) __nonnull ((1)) __wur;
 #endif
 
 #if (defined _GNU_SOURCE && defined __UCLIBC_HAS_THREADS_NATIVE__)
-pid_t gettid(void);
+extern __pid_t gettid(void);
 #endif
 
 __END_DECLS
-- 
2.47.3

