diff -Nur uClibc-ng-1.0.56.orig/extra/Configs/Config.sparc uClibc-ng-1.0.56/extra/Configs/Config.sparc
--- uClibc-ng-1.0.56.orig/extra/Configs/Config.sparc	2026-01-10 12:09:58.000000000 +0100
+++ uClibc-ng-1.0.56/extra/Configs/Config.sparc	2026-02-13 15:01:56.035360315 +0100
@@ -27,4 +27,7 @@
 config CONFIG_SPARC_V8
 	bool "SPARC v8"
 
+config CONFIG_SPARC_LEON3
+	bool "SPARC LEON3"
+
 endchoice
diff -Nur uClibc-ng-1.0.56.orig/Rules.mak uClibc-ng-1.0.56/Rules.mak
--- uClibc-ng-1.0.56.orig/Rules.mak	2026-01-10 12:09:58.000000000 +0100
+++ uClibc-ng-1.0.56/Rules.mak	2026-02-13 15:02:49.763525192 +0100
@@ -383,6 +383,7 @@
 ifeq ($(TARGET_ARCH),sparc)
 	CPU_CFLAGS-$(CONFIG_SPARC_V7)+=-mcpu=v7
 	CPU_CFLAGS-$(CONFIG_SPARC_V8)+=-mcpu=v8
+	CPU_CFLAGS-$(CONFIG_SPARC_LEON3)+=-mcpu=leon3
 	CPU_CFLAGS-$(CONFIG_SPARC_V9)+=-mcpu=v9
 	CPU_CFLAGS-$(CONFIG_SPARC_V9B)+=$(call check_gcc,-mcpu=v9b,-mcpu=ultrasparc)
 endif
