diff -Nur gcc-4.4.7.orig/gcc/toplev.h gcc-4.4.7/gcc/toplev.h
--- gcc-4.4.7.orig/gcc/toplev.h	2009-02-20 16:20:38.000000000 +0100
+++ gcc-4.4.7/gcc/toplev.h	2017-06-30 21:00:00.981033517 +0200
@@ -167,38 +167,6 @@
 extern bool fast_math_flags_set_p	(void);
 extern bool fast_math_flags_struct_set_p (struct cl_optimization *);
 
-/* Return log2, or -1 if not exact.  */
-extern int exact_log2                  (unsigned HOST_WIDE_INT);
-
-/* Return floor of log2, with -1 for zero.  */
-extern int floor_log2                  (unsigned HOST_WIDE_INT);
-
-/* Inline versions of the above for speed.  */
-#if GCC_VERSION >= 3004
-# if HOST_BITS_PER_WIDE_INT == HOST_BITS_PER_LONG
-#  define CLZ_HWI __builtin_clzl
-#  define CTZ_HWI __builtin_ctzl
-# elif HOST_BITS_PER_WIDE_INT == HOST_BITS_PER_LONGLONG
-#  define CLZ_HWI __builtin_clzll
-#  define CTZ_HWI __builtin_ctzll
-# else
-#  define CLZ_HWI __builtin_clz
-#  define CTZ_HWI __builtin_ctz
-# endif
-
-extern inline int
-floor_log2 (unsigned HOST_WIDE_INT x)
-{
-  return x ? HOST_BITS_PER_WIDE_INT - 1 - (int) CLZ_HWI (x) : -1;
-}
-
-extern inline int
-exact_log2 (unsigned HOST_WIDE_INT x)
-{
-  return x == (x & -x) && x ? (int) CTZ_HWI (x) : -1;
-}
-#endif /* GCC_VERSION >= 3004 */
-
 /* Functions used to get and set GCC's notion of in what directory
    compilation was started.  */
 
diff -Nur gcc-4.4.7.orig/libiberty/md5.c gcc-4.4.7/libiberty/md5.c
--- gcc-4.4.7.orig/libiberty/md5.c	2008-03-11 15:08:53.000000000 +0100
+++ gcc-4.4.7/libiberty/md5.c	2026-08-27 03:22:17.849165305 +0200
@@ -26,15 +26,8 @@
 #endif
 
 #include <sys/types.h>
-
-#if STDC_HEADERS || defined _LIBC
-# include <stdlib.h>
-# include <string.h>
-#else
-# ifndef HAVE_MEMCPY
-#  define memcpy(d, s, n) bcopy ((s), (d), (n))
-# endif
-#endif
+#include <stdlib.h>
+#include <string.h>
 
 #include "ansidecl.h"
 #include "md5.h"
diff -Nur gcc-4.4.7.orig/libiberty/regex.c gcc-4.4.7/libiberty/regex.c
--- gcc-4.4.7.orig/libiberty/regex.c	2005-07-22 05:55:44.000000000 +0200
+++ gcc-4.4.7/libiberty/regex.c	2026-08-27 03:22:00.273088220 +0200
@@ -124,12 +124,7 @@
    even if config.h says that we can.  */
 #  undef REL_ALLOC
 
-#  if defined STDC_HEADERS || defined _LIBC
-#   include <stdlib.h>
-#  else
-char *malloc ();
-char *realloc ();
-#  endif
+#  include <stdlib.h>
 
 /* When used in Emacs's lib-src, we need to get bzero and bcopy somehow.
    If nothing else has been done, use the method below.  */
