diff -Nur linux-4.14.334.orig/arch/mips/rb532/devices.c linux-4.14.334/arch/mips/rb532/devices.c
--- linux-4.14.334.orig/arch/mips/rb532/devices.c	2023-12-20 15:32:39.000000000 +0100
+++ linux-4.14.334/arch/mips/rb532/devices.c	2026-04-21 10:44:49.676097062 +0200
@@ -225,11 +225,11 @@
 static struct plat_serial8250_port rb532_uart_res[] = {
 	{
 		.type           = PORT_16550A,
-		.membase	= (char *)KSEG1ADDR(REGBASE + UART0BASE),
+		.mapbase        = REGBASE + UART0BASE,
 		.irq		= UART0_IRQ,
 		.regshift	= 2,
 		.iotype		= UPIO_MEM,
-		.flags		= UPF_BOOT_AUTOCONF,
+		.flags		= UPF_BOOT_AUTOCONF | UPF_IOREMAP,
 	},
 	{
 		.flags		= 0,
