diff -Nur linux-6.1.44.orig/arch/m68k/fpsp040/skeleton.S linux-6.1.44/arch/m68k/fpsp040/skeleton.S
--- linux-6.1.44.orig/arch/m68k/fpsp040/skeleton.S	2023-08-08 20:03:51.000000000 +0200
+++ linux-6.1.44/arch/m68k/fpsp040/skeleton.S	2023-08-14 04:47:03.786841535 +0200
@@ -499,13 +499,13 @@
 	dbf	%d0,morein
 	rts
 
-	.section .fixup,#alloc,#execinstr
+	.section .fixup,"ax"
 	.even
 1:
 	jbsr	fpsp040_die
 	jbra	.Lnotkern
 
-	.section __ex_table,#alloc
+	.section __ex_table,"a"
 	.align	4
 
 	.long	in_ea,1b
diff -Nur linux-6.1.44.orig/arch/m68k/ifpsp060/os.S linux-6.1.44/arch/m68k/ifpsp060/os.S
--- linux-6.1.44.orig/arch/m68k/ifpsp060/os.S	2023-08-08 20:03:51.000000000 +0200
+++ linux-6.1.44/arch/m68k/ifpsp060/os.S	2023-08-14 04:47:03.798841249 +0200
@@ -379,11 +379,11 @@
 
 
 | Execption handling for movs access to illegal memory
-	.section .fixup,#alloc,#execinstr
+	.section .fixup,"ax"
 	.even
 1:	moveq		#-1,%d1
 	rts
-.section __ex_table,#alloc
+.section __ex_table,"a"
 	.align 4
 	.long	dmrbuae,1b
 	.long	dmrwuae,1b
diff -Nur linux-6.1.44.orig/arch/m68k/kernel/relocate_kernel.S linux-6.1.44/arch/m68k/kernel/relocate_kernel.S
--- linux-6.1.44.orig/arch/m68k/kernel/relocate_kernel.S	2023-08-08 20:03:51.000000000 +0200
+++ linux-6.1.44/arch/m68k/kernel/relocate_kernel.S	2023-08-14 04:47:03.802841154 +0200
@@ -26,7 +26,7 @@
 	lea %pc@(.Lcopy),%a4
 2:	addl #0x00000000,%a4		/* virt_to_phys() */
 
-	.section ".m68k_fixup","aw"
+	.section .m68k_fixup,"aw"
 	.long M68K_FIXUP_MEMOFFSET, 2b+2
 	.previous
 
@@ -49,7 +49,7 @@
 	lea %pc@(.Lcont040),%a4
 5:	addl #0x00000000,%a4		/* virt_to_phys() */
 
-	.section ".m68k_fixup","aw"
+	.section .m68k_fixup,"aw"
 	.long M68K_FIXUP_MEMOFFSET, 5b+2
 	.previous
 
