diff -Nur linux-6.12.25.orig/arch/mips/vdso/Makefile linux-6.12.25/arch/mips/vdso/Makefile
--- linux-6.12.25.orig/arch/mips/vdso/Makefile	2025-04-25 10:48:08.000000000 +0200
+++ linux-6.12.25/arch/mips/vdso/Makefile	2025-04-30 19:42:08.538362252 +0200
@@ -30,7 +30,7 @@
 	-O3 -g -fPIC -fno-strict-aliasing -fno-common -fno-builtin -G 0 \
 	-mrelax-pic-calls $(call cc-option, -mexplicit-relocs) \
 	-fno-stack-protector -fno-jump-tables -DDISABLE_BRANCH_PROFILING \
-	$(call cc-option, -fno-asynchronous-unwind-tables)
+	$(call cc-option, -fno-asynchronous-unwind-tables) -std=gnu11
 aflags-vdso := $(ccflags-vdso) \
 	-D__ASSEMBLY__ -Wa,-gdwarf-2
 
