diff -Nur linux-4.19.320.orig/arch/mips/vdso/Makefile linux-4.19.320/arch/mips/vdso/Makefile
--- linux-4.19.320.orig/arch/mips/vdso/Makefile	2024-08-19 05:32:18.000000000 +0200
+++ linux-4.19.320/arch/mips/vdso/Makefile	2026-04-19 09:15:00.183108904 +0200
@@ -18,7 +18,7 @@
 	-O2 -g -fPIC -fno-strict-aliasing -fno-common -fno-builtin -G 0 \
 	-DDISABLE_BRANCH_PROFILING \
 	$(call cc-option, -fno-asynchronous-unwind-tables) \
-	$(call cc-option, -fno-stack-protector)
+	$(call cc-option, -fno-stack-protector) -std=gnu11
 aflags-vdso := $(ccflags-vdso) \
 	-D__ASSEMBLY__ -Wa,-gdwarf-2
 
